### ISSUE TITLE
Union of Overloaded Descriptor Types Leads False Positive Error

### ISSUE THREAD

**Issue by mfish33** [microsoft/pyright#10255]
**Describe the bug**
This is a simplification of a real-world example where descriptors describe the differing behavior of assignments when a type is used in a C++ extension vs a Python context. I think the solution to this problem would be to perform a union of the matched descriptor overload. In the below example, this would make the "effective" descriptor for `foo` of type `(instance: Any, Value: A | B) -> None`. 

Note: mypy does not report this as an error

**Code**

```python
# pyright: strict

from __future__ import annotations

from typing import overload, cast, Any
from typing_extensions import Self

class Magic:
   ...

class A:
   @overload
   def __set__( self, instance: Magic, value: Self | str ) -> None:
      ...
   @overload
   def __set__( self, instance: Any, value: Self ) -> None:
      ...
   def __set__( self, instance: Any, value: Any ) -> None:
      ...

   @overload
   def __get__( self, instance: Magic, owner: Any = None ) -> str:
      return cast( str, None )
   @overload
   def __get__( self, instance: Any, owner: Any = None ) -> Self:
      return cast( Self, None )
   def __get__( self, instance: Any, owner: Any = None ) -> Any:
      return cast( Any, None )

class B:
   @overload
   def __set__( self, instance: Magic, value: Self | int ) -> None:
      ...
   @overload
   def __set__( self, instance: Any, value: Self ) -> None:
      ...
   def __set__( self, instance: Any, value: Any ) -> None:
      ...

   @overload
   def __get__( self, instance: Magic, owner: Any = None ) -> int:
      return cast( int, None )
   @overload
   def __get__( self, instance: Any, owner: Any = None ) -> Self:
      return cast( Self, None )
   def __get__( self, instance: Any, owner: Any = None ) -> Any:
      return cast( Any, None )

class PythonContainer:
   foo: A | B

   def __init__( self, foo: A | B ) -> None:
      self.foo = foo
```
**Command-line**
Pyright version 1.1.398
```
➜  TtBaseTest git:(ttSingletonRevert.0) ✗ pyright test.py       
/src/TtBaseTest/test.py
  /src/TtBaseTest/test.py:53:12 - error: Cannot assign to attribute "foo" for class "MagicContainer*"
    No overloads for "__set__" match the provided arguments
    Argument of type "A | B" cannot be assigned to parameter "value" of type "A" in function "__set__"
      Type "A | B" is not assignable to type "A"
        "B" is not assignable to "A"
    No overloads for "__set__" match the provided arguments
    Argument of type "A | B" cannot be assigned to parameter "value" of type "B" in function "__set__"
      Type "A | B" is not assignable to type "B"
        "A" is not assignable to "B" (reportAttributeAccessIssue)
1 error, 0 warnings, 0 informations 
```

**Comment by erictraut** [comment]
Pyright is working correctly here, so this isn't a bug. The solution you're suggesting would be unsound from a typing standpoint.

The attribute `foo` is declared in the body of the class, so it is considered a [regular class variable](https://microsoft.github.io/pyright/#/type-concepts-advanced?id=regular-class-variable) by pyright.

The type of the attribute `foo` is a union of `A` and `B`. The type of the parameter `foo` is the same. There is no constraint here that says that they must both be either `A` or `B`. 

If you want pyright to interpret the attribute `foo` as a [pure instance variable](https://microsoft.github.io/pyright/#/type-concepts-advanced?id=pure-instance-variables) — one that can be set only on instances of the class rather than the class itself, then you would need to declare it like this:

```python
class PythonContainer:
    def __init__(self, foo: A | B) -> None:
        self.foo: A | B = foo
```

Or you can omit the annotation, and pyright will infer its type as `A | B`.

**Comment by mfish33** [comment]
Wow! That was fast. A pure instance variable works for us. 

Thanks for taking the time to respond. Have a nice weekend!
